src/modules/Elsa.Resilience.Core/Services/ResilienceStrategyCatalog.cs

```
@@ -0,0 +1,31 @@
+namespace Elsa.Resilience;
+
+public class ResilienceStrategyCatalog : IResilienceStrategyCatalog
+{
+    private readonly Lazy<Task<IEnumerable<IResilienceStrategy>>> _strategies;
+    private readonly IEnumerable<IResilienceStrategySource> _providers;
+
+    public ResilienceStrategyCatalog(IEnumerable<IResilienceStrategySource> providers)
+    {
+        _providers = providers;
+        _strategies = new(GetStrategiesInternalAsync, LazyThreadSafetyMode.ExecutionAndPublication);
+    }
+    
+    public Task<IEnumerable<IResilienceStrategy>> ListAsync(CancellationToken cancellationToken = default)
+    {
+        return _strategies.Value;
+    }
+
+    public async Task<IResilienceStrategy?> GetAsync(string id, CancellationToken cancellationToken = default)
+    {
+        var strategies = await ListAsync(cancellationToken);
+        return strategies.FirstOrDefault(x => x.Id == id);
+    }
+
+    private async Task<IEnumerable<IResilienceStrategy>> GetStrategiesInternalAsync()
+    {
+        var strategies = new List<IResilienceStrategy>();
+        foreach (var provider in _providers) strategies.AddRange(await provider.GetStrategiesAsync());
```

How do we want to handle Strategies with the same ID or name for from either the same or different providers?

---

Nice catch. I think that either we leave this up to the user, or we prefix IDs with e.g. the type name of the resilience strategy source, hashed or otherwise. Tricky part with that might be that we must ensure the type names never get renamed, or else.

Another idea might be to use an attribute that uniquely identifies the source type. Obviously, this still doesn't guarantee uniqueness if two independent developers happen to pick the same identifier, but given those small odds, at least we could throw an exception.

Finally, if we did want to ensure developers won't have to deal with each other in this manner, we could put the burden with the application developer and require them to manually register the sources, where they should then also provide some sort of prefix. I'm not a fan of this because it makes it more complicated for the application developer.

What do you think might be the best solution?

---

I opted for the second approach where we can control the prefix of the identities using a attribute. This prevents us from renaming types without realizing we'd be breaking people's workflows without requiring application developers to explicitly provide a prefix or having to be concerned with unique identities across strategy sources.

Let me know what you think.